client/lib/analytics/signup.js

```diff
@@ -0,0 +1,94 @@
+/**
+ * Internal dependencies
+ */
+import analytics from 'lib/analytics';
+import {
+	adTrackSignupStart,
+	adTrackSignupComplete,
+	adTrackRegistration,
+} from 'lib/analytics/ad-tracking';
+
+export function recordSignupStart( flow, ref ) {
+	// Tracks
+	analytics.tracks.recordEvent( 'calypso_signup_start', { flow, ref } );
+	// Google Analytics
+	analytics.ga.recordEvent( 'Signup', 'calypso_signup_start' );
+	// Marketing
+	adTrackSignupStart( flow );
+}
+
+export function recordSignupComplete(
+	{ flow, isNewUser, isNewSite, hasCartItems, isNew7DUserSite },
+	now
+) {
+	if ( ! now ) {
+		// Delay using the analytics localStorage queue.
+		return analytics.queue.add(
```

Comment:
@jsnajdr I believe the `analytics.queue` object expects the passed function name to be part of the `analytics` object and since it is not the case anymore, the signup tracking appears broken for me, ie I don't see it being fired.

For example after completing the signup process I don't see any reference to `calypso_signup_complete` in my browser's Network activity.

The initial signup part seems ok both the cookie banner and the gdpr logic seem to work as expected. To fully test all code paths one needs to be connected through a GDPR country and make sure the cookie banner appears and then click it.

This is what I used to enable all relevant features and see the logs:
```
document.cookie = 'flags=gdpr-banner,google-analytics,ad-tracking; path=/; domain=.' + document.location.hostname.split( '.' ).slice( -2 ).join( '.' );
localStorage.setItem( 'debug', 'calypso:analytics:*' );
```

Reply:
That's a great catch @MicBosi :+1: I didn't know what `analytics.queue` does until now.

I fixed this bug by importing `recordSignupComplete` from the separate module and adding a property to `analytics`. That defeats the attempt to improve code splitting, but I don't see any better solution right now.

The queueing means that `recordSignupComplete` is not called only in the Signup section, but potentially also anywhere else -- on the first page view after loading, whatever page that view is.

@jaswrks Do I understand correctly that `analytics.queue` prevents cancelling the `recordSignupComplete` async task when leaving the current page? It needs to wait for the tracker scripts to load etc.

How does Google Tag Manager handle such a case? It looks like something that's common and happens all the time: sending analytics correctly and completely before leaving a page.

Reply:
@jsnajdr Thanks for fixing that.

> Do I understand correctly that analytics.queue prevents cancelling the `recordSignupComplete` async task when leaving the current page? It needs to wait for the tracker scripts to load etc.

That sounds about right to me and for the Signup Complete event in particular it's needed because we do a full reload of the page which doesn't usually happen. As far as I know waiting for the trackers to load hasn't been that big of an issue, while the page reload we do right after signup complete is.

To be honest I'm not sure how much better our tracking is with the queue mechanism, that is, it probably is but I don't have hard numbers on the topic. I also wish we didn't have to use it so if we figured out a way to not need it I think we would all be happy about it.

> How does Google Tag Manager handle such a case? It looks like something that's common and happens all the time: sending analytics correctly and completely before leaving a page.

Yes it is a common problem in general although not so much in Calypso since most page views are "virtual" page views, we don't actually reload the page, while for landing pages for example this is a real problem for things like click tracking.

In theory the Beacon API has been introduced to address this very issue but it needs to be supported by any given tracker. Google Analytics does for example. Many others don't, including Tacks. Google Tag Manager unfortunately does not affect this issue at all since supporting the Beacon API depends entirely on the tracker being fired within GTM. So from this particular point of view GTM won't make things neither better or worse as it will be triggered from within the `recordSignupComplete` function.

Reply:
> In theory the Beacon API has been introduced to address this very issue but it needs to be supported by any given tracker.

Also, the Beacon API is useful only when you want to send a single XHR request and want a guarantee that it will finish even if the page gets unloaded at the same time.

A `recordSignupComplete` call is potentially much complex than that. It may send a geolocation request to the REST API to figure out if it's in a GDPR country, then load several tracking scripts, and only then send the actual beacon tracking requests. The page can be long gone before that beacon requests even start.

In any case, @MicBosi please let me know if this PR can be merged now. #39001 depends on the changes here.